forgerock-authenticator/src/main/java/org/forgerock/android/auth/biometric/BiometricAuth.java

```diff
@@ -0,0 +1,217 @@
+     */
+    public FragmentActivity getActivity() {
+        return activity;
+    }
+
+    /*
+     * Starts authentication process.
+     */
+    public void authenticate() {
+        // API 29 and above
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+            if(hasBiometricCapability()) {
+                initBiometricAuthentication();
+            }
+            return;
+        }
+
+        // API 23 - 28 with fingerprint
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && fingerprintManager != null) {
+            if (fingerprintManager.hasEnrolledFingerprints()) {
+                initBiometricAuthentication();
+            }
+            return;
+        }
+
+        // API 23 or higher with just device credentials
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M
+                && this.keyguardManager != null
+                && this.keyguardManager.isDeviceSecure()) {
+            initDeviceCredentialAuthentication();
+        }
+    }
+
+    private void setServicesFromActivity(@NonNull FragmentActivity activity) {
+        Context context = activity.getBaseContext();
+        this.biometricManager = BiometricManager.from(activity);
+        this.keyguardManager = (KeyguardManager) context.getSystemService(KEYGUARD_SERVICE);
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            this.fingerprintManager = (FingerprintManager) context.getSystemService(Context.FINGERPRINT_SERVICE);
+        }
+    }
+
+    private BiometricPrompt initBiometricPrompt() {
+        Executor executor = ContextCompat.getMainExecutor(this.activity);
+        BiometricPrompt biometricPrompt = new BiometricPrompt(
+                this.activity, executor, new BiometricPrompt.AuthenticationCallback() {
+            @Override
+            public void onAuthenticationError(int errorCode,
+                                              @NonNull CharSequence errString) {
+                super.onAuthenticationError(errorCode, errString);
+                biometricAuthListener.onError(errorCode, errString.toString());
+            }
+
+            @Override
+            public void onAuthenticationSucceeded(
+                    @NonNull BiometricPrompt.AuthenticationResult result) {
+                super.onAuthenticationSucceeded(result);
+                biometricAuthListener.onSuccess(result);
+            }
+
+            @Override
+            public void onAuthenticationFailed() {
+                super.onAuthenticationFailed();
+            }
+        });
+
+        int authenticators = BIOMETRIC_STRONG;
+        if(this.allowDeviceCredentials) {
+            authenticators += DEVICE_CREDENTIAL;
+        }
+
+        BiometricPrompt.PromptInfo.Builder builder = new BiometricPrompt.PromptInfo.Builder()
+                .setAllowedAuthenticators(authenticators);
+        if(this.title == null) {
+            builder.setTitle("Biometric Authentication for login");
```

Comment:
Should setTitle, subtitle, and negative button to builder when not null?

Reply:
yes, totally missed it